File: LeetCode/30-Day-LeetCoding-Challenge-master/Min_Stack.cpp
class MinStack {
public:
    /** initialize your data structure here. */
    vector<int> stack, min;
    MinStack() { } // constructor
    
    void push(int x) {
        stack.push_back(x);
        if(min.empty() or min.back() >= x)
            min.push_back(x);
    }
    
    void pop() {
        if(stack.back() == min.back()) {
            min.pop_back();
            stack.pop_back();
        }
        else 
            stack.pop_back();
    }
    
    int top() {
        if(stack.size() != 0) 
            return stack.back();
        else 
            return NULL;
    }
    
    int getMin() {
        if(min.size() != 0) 
           return min.back();
        else 
            return NULL;
    }
};

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack* obj = new MinStack();
 * obj->push(x);
 * obj->pop();
 * int param_3 = obj->top();
 * int param_4 = obj->getMin();
 */

File: LeetCode/30-Day-LeetCoding-Challenge-master/Min_Stack.py
class MinStack:

    def __init__(self):
        """
        initialize your data structure here.
        """
        self.stack = []
        self.min = []

    def push(self, x: int) -> None:
        self.stack.append(x)
        if len(self.min) == 0 or self.min[-1] >= x:
            self.min.append(x)
        
        
    def pop(self) -> None:
        if len(self.stack) != 0 and len(self.min) != 0 and self.stack[-1] == self.min[-1]:
            self.min.pop()
            self.stack.pop()
        else:
            self.stack.pop()
        

    def top(self) -> int:
        if len(self.stack) != 0:
            return self.stack[-1]
        

    def getMin(self) -> int:
        if len(self.min) != 0:
            return self.min[-1]
        


# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(x)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()
